docs/auth/sexpr.md

```diff
@@ -0,0 +1,563 @@
+  F+ service and returns the result. (Scalar)
+
+* `map` (*): `["map" [bind ...body] ...list]`. Map starts by
+  list-evaluating `list`, using the current bindings. Then, for each
+  element in the result, a new set of bindings is created with that
+  element bound to `bind`. The `body` is list-evaluated against each set
+  of bindings and the results concatenated and returned. (List)
+
+* `merge`: `["merge" ...objs]`. Merge the properties of all objects,
+  with values from later objects overwriting those from earlier objects.
+  (Scalar)
+
+* `principal`: `["principal"]`. Returns the UUID of the principal to
+  whom this ACE applies. (Scalar)
+
+* `quote` (*): `["quote" value]`. Returns its value unchanged and
+  unevaluated. (Scalar)
+
+### Template definitions
+
+A permission template is a function which accepts zero or more arguments
+and returns a list of permission grants. When ACLs are looked up all
+permission templates are expanded recursively and the result included in
+the list of ACEs returned. Some form of recursion control will obviously
+be necessary.
+
+If a template is called with too many arguments additional arguments are
+ignored. If a template is called with too few arguments the additional
+parameters are bound to `null`.
+
+A template definition is represented in JSON as an S-expression built
+from lists (i.e. arrays) and strings. The top level of the definition
+must be a list of the form `[[...params] ...results]`. The first
+element is a list of parameter names; these must be strings, and when
+the template is called the arguments will be bound to the parameters as
+though via `let`. The remainder of the definition forms the result list;
+each item will be evaluated and the result flattened into the template's
+return list.
+
+## Examples
+
+Notation:
+
+    ∈           Class membership
+    ⊂           Class subset
+    →           Permission template definition
+    u:[]        Permission grant
+    u:<t i>     Identity
+    ⇒           Permission expansion
+
+Titlecase strings represent UUIDs. Lowercase strings represent JSON
+strings. JSON arrays are represented without commas. JSON objects allow
+trailing commas. For example, this
+
+    [ReadConfig {app: Address, obj: ConfigDB}]
+
+represents this JSON
+
+    [
+      "4a339562-cd57-408d-9d1a-6529a383ea4b",
+      {   
+        "app": "8e32801b-f35a-4cbf-a5c3-2af64d3debd7",
+        "obj": "36861e8d-9152-40c4-8f08-f51c2d7e3c25"
+      }
+    ]
+
+Each example builds on the others. Comments start with `#`. Object
+contents are abbreviated where there is a lot of repetition.
+
+### Sparkplug Node publishing
+
+    # Template definitions
+    ReadOwnConfig → [[app]
+      [ReadConfig: { app: [app], obj: [principal] }]]
+
+    SpTopic → [[type addr]
+      [if [has [addr] device] 
+        [format "spBv1.0/%s/N%s/%s" [addr group] [type] [addr node]]
+        [format "spBv1.0/%s/D%s/%s" 
+          [addr group] [type] [addr node] [addr device]]]]
```

Comment:
Why does this have four arguments but only three `%s`?

Reply:
Looking at this again, it's completely wrong. The correct version is here: https://github.com/AMRC-FactoryPlus/amrc-connectivity-stack/blob/5774de3a7f56ba603689b3cec970cbc59405124a/acs-auth/bin/sexpr.js#L218